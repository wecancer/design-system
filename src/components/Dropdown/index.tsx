import React, {useState, useRef} from 'react'
import styled, {css} from 'styled-components'

import OutsideEvent from '../OutsideEvent'

type AxisX = 'left' | 'right'

const Wrapper = styled(OutsideEvent)`
  position: relative;
  display: inline-block;
`

const TriggerContainer = styled.div``

const BoxContainer = styled.div<{triggerHeight: number; axisX: AxisX}>`
  ${({triggerHeight, axisX}) => css`
    position: absolute;
    top: ${triggerHeight + 10}px;
    ${axisX === 'left'
      ? css`
          left: 0;
        `
      : css`
          right: 0;
        `}
  `}
`

export type ActionParams = {
  isOpen: boolean
  handleClose(): void
  handleToggle(): void
}

export type Props = {
  axisX?: AxisX
  trigger(params: ActionParams): React.ReactNode
  children(params: ActionParams): React.ReactNode
  className?: string
}

const Dropdown = ({axisX = 'left', children, trigger, ...props}: Props) => {
  const refTrigger = useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = useState(false)
  const handleClose = () => setOpen(false)
  const handleToggle = () => setOpen(!isOpen)

  return (
    <Wrapper onClickOutside={handleClose} {...props}>
      <TriggerContainer ref={refTrigger}>
        {trigger({isOpen, handleClose, handleToggle})}
      </TriggerContainer>
      {isOpen && (
        <BoxContainer
          axisX={axisX}
          triggerHeight={refTrigger.current?.offsetHeight || 0}
        >
          {children({isOpen, handleClose, handleToggle})}
        </BoxContainer>
      )}
    </Wrapper>
  )
}

export default Dropdown
