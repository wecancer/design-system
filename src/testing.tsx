import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
