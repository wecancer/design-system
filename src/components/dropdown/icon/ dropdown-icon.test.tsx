import Dropdown from '.'
import { render, fireEvent } from '../../../testing'

describe('<DropdownIcon />', () => {
  const options = [
    {
      label: 'item 1',
      id: '01',
    },
    {
      label: 'item 2',
      id: '02',
    },
    {
      label: 'item 3',
      id: '03',
    },
  ]

  it('should render snapshot correctly', () => {
    const { container } = render(
      <>
        <Dropdown icon="more" items={[]} />
        <Dropdown icon="more" items={options} />
      </>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should open the options when the lavel is clicked', () => {
    const { getByRole, container } = render(
      <Dropdown icon="moreVert" items={options} />,
    )
    const icon = container.querySelector('*[data-icon-name="moreVert"]')

    if (!icon) throw new Error(`Can't find the button with the text "my label"`)

    fireEvent.click(icon)
    expect(getByRole('listbox')).toBeInstanceOf(HTMLElement)
  })
})
