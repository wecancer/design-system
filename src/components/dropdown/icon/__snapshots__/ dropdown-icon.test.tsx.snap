// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownIcon /> should render snapshot correctly 1`] = `
<div
  class="sc-dkzDqf hUMDeT"
>
  <div
    class="sc-hKMtZM"
  >
    <button
      class="sc-gsnTZi gxeoBA"
    >
      <div
        data-icon-name="more"
        style="display: inline-flex; align-items: center; justify-content: center;"
      >
        <svg
          fill="none"
          height="16px"
          viewBox="0 0 24 24"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="7"
            cy="12"
            fill="currentColor"
            r="2"
            transform="rotate(-90 7 12)"
          />
          <circle
            cx="17"
            cy="12"
            fill="currentColor"
            r="2"
            transform="rotate(-90 17 12)"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
